### ISSUE TITLE
Batch norm internal variables keep changing during inference

### ISSUE THREAD

**Issue by ZeliangSu** [tensorflow/tensorflow#35167]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 16.04
- Mobile device if the issue happens on mobile device: Dell Precision Tower 7910
- TensorFlow installed from (source or binary): tensorflow-gpu 1.14.0
- TensorFlow version (use command below): tensorflow-gpu 1.14.0
- Installed from: conda
- Python version: python 3.6
- GCC/Compiler version (if compiling from source): 6.5.0
- CUDA/cuDNN version: 10.1 but nvcc 7.5
- GPU model and memory:Quadro M4000

**Describe the current behavior**
I've encountered weird behavior of batch normalization in TF1.14 (I will update ulteriorly for TF2). The accuracy decrease a lot after the graph freezing. I identified that the origin of this discrepency is due to the batch normalization. A smaller model code to reproduce this behavior is attached at the end of this post. 
After training, I freeze the graph with `tf.graph_util.convert_variables_to_constants()` then optimize with `optimize_for_inference()`. From the results below, you can see that the from training to testing it changed slightly but **just after** freezing, the **moving_avg, beta and gamma changed tremendously** but not the moving_std:

> - [x] Training:   

>0%|          | 0/100 [00:00<=X, X/its] mov_avg: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.],  mov_std: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.],  beta: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.],  gamma: [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
> 
> 100%|██████████| 100/100 [00:01<00:00, 79.00it/s] mov_avg:
> [**-0.18366139  0.24028867  0.91940075 -0.07153109**  0.08811506
> -0.5552308
>   0.12759416  0.11020644  0.47183645 -0.91372997 -0.35405988  0.5768641 ],  mov_std: [**0.36972958 0.36972958 0.36972958 0.36972958** 0.36972958 0.36972958
>  0.36972958 0.36972958 0.36972958 0.36972958 0.36972958 0.36972958],  beta: [**0.         0.         0.         0.00995897** 0.         0.
>  0.00995881 0.00995875 0.         0.00989968 0.         0.        ],  gamma: [**1.         1.         1.         1.0009596**  1.         1.
>  **0.99905616 0.99986655 1.         1.0005279**  1.         1.        ]
> 
> - [x] Testing: 

>0%|          | 0/5 [00:00<=X, Xit/s] mov_avg: [**-0.18473879  0.24169827  0.92479414 -0.07199407**  0.08863196
> -0.5584879
>   0.12828277  0.11082216  0.47460434 -0.9191201  -0.35613686  0.5802481 ],  mov_std: [**0.36603227 0.36603227 0.36603227 0.36603227** 0.36603227 0.36603227
>  0.36603227 0.36603227 0.36603227 0.36603227 0.36603227 0.36603227],  beta: [**0.         0.         0.         0.01005934** 0.         0.
>  0.01005919 0.01005913 0.         0.00999967 0.         0.        ],  gamma: [**1.         1.         1.         1.0009888**  1.         1.
>  0.99904144 0.999887   1.         1.0005565  1.         1.        ]
> 
> 100%|██████████| 100/100 [00:00<00:00, 182.68it/s] mov_avg:
> [**-0.18473879  0.24169827  0.92479414 -0.07199407  0.08863196**
> -0.5584879
>   0.12828277  0.11082216  0.47460434 -0.9191201  -0.35613686  0.5802481 ],  mov_std: [**0.36603227 0.36603227 0.36603227 0.36603227 0.36603227** 0.36603227
>  0.36603227 0.36603227 0.36603227 0.36603227 0.36603227 0.36603227],  beta: [**0.         0.         0.         0.01005934 0.**         0.
>  0.01005919 0.01005913 0.         0.00999967 0.         0.        ],  gamma: [**1.         1.         1.         1.0009888  1.**         1.
>  0.99904144 0.999887   1.         1.0005565  1.         1.        ]

> - [x] Optimized (is_training = False) 

>0%|          | 0/100 [00:00<=X, Xit/s] mov_avg: [ **0.31595632 -0.19200048 -0.80097497  0.22726376**
> **-0.2556441**  -0.1720414
>   0.05251308  0.5521817  -0.93566465 -0.14799471 -0.11172786 -0.03040024],  mov_std: [**0.36603227 0.36603227 0.36603227 0.36603227 0.36603227** 0.36603227
>  0.36603227 0.36603227 0.36603227 0.36603227 0.36603227 0.36603227],  beta: [**0.         0.         0.         0.         0.00999963** 0.
>  0.01005759 0.         0.         0.         0.         0.        ],  gamma: [**1.         1.         1.         1.         1.0017155**  1.
>  0.99796325 1.         1.         1.         1.         1.        ]
> 
> 100%|██████████| 100/100 [00:00<00:00, 404.37it/s] mov_avg: [
> 0.31595632 -0.19200048 -0.80097497  0.22726376 -0.2556441  -0.1720414
>   0.05251308  0.5521817  -0.93566465 -0.14799471 -0.11172786 -0.03040024],  mov_std: [0.36603227 0.36603227 0.36603227 0.36603227 0.36603227 0.36603227
>  0.36603227 0.36603227 0.36603227 0.36603227 0.36603227 0.36603227],  beta: [0.         0.         0.         0.         0.00999963 0.
>  0.01005759 0.         0.         0.         0.         0.        ],  gamma: [1.         1.         1.         1.         1.0017155  1.
>  0.99796325 1.         1.         1.         1.         1.        ]

**Describe the expected behavior**
I was expecting these local variables **not change when we switch the training parameter between True/False or while we freeze it**.

**Code to reproduce the issue**

```
import tensorflow as tf
from tqdm import tqdm
import numpy as np
from util import print_nodes_name_shape, check_N_mkdir
from tensorflow.python.tools.optimize_for_inference_lib import optimize_for_inference
from tensorflow.python.framework import dtypes
import os


input_ph = tf.placeholder(tf.float32, shape=(None, 2, 2, 1), name='input_ph')
output_ph = tf.placeholder(tf.float32, shape=(None, 2, 2, 3), name='output_ph')
is_training = tf.placeholder(tf.bool, shape=[], name='is_training')

# build a one layer Full layer with BN and save 2 ckpts
with tf.name_scope('model'):
    out = tf.reshape(input_ph, shape=(-1, 2 * 2 * 1), name='flatten')
    with tf.variable_scope('dnn1', reuse=False):
        w1 = tf.get_variable('w1', dtype=tf.float32, shape=[4 * 1, 4 * 3], initializer=tf.initializers.glorot_normal())
        # b1 = tf.get_variable('b1', dtype=tf.float32, shape=[4 * 3], initializer=tf.initializers.glorot_normal())
    # out = tf.matmul(out, w1) + b1
    out = tf.matmul(out, w1)
    out = tf.layers.batch_normalization(out, training=is_training, name='BN')
    logits = tf.nn.relu(out)
    logits = tf.reshape(logits, shape=(-1, 2, 2, 3))

with tf.name_scope('loss'):
    MSE = tf.losses.mean_squared_error(labels=output_ph, predictions=logits)

with tf.name_scope('operations'):
    opt = tf.train.AdamOptimizer(learning_rate=0.0001, name='Adam')
    grads = opt.compute_gradients(MSE)
    train_op = opt.apply_gradients(grads, name='apply_grad')

# train
with tf.Session() as sess:
    # prepare
    graph = tf.get_default_graph()
    print_nodes_name_shape(graph)
    saver = tf.train.Saver()

    # init variables
    sess.run([tf.global_variables_initializer(), tf.local_variables_initializer()])
    update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
    print(update_ops)  # note: [<tf.Operation 'model/BN/cond_2/Merge' type=Merge>, <tf.Operation 'model/BN/cond_3/Merge' type=Merge>]
    saver.save(sess, './dummy/ckpt/step0')
    # train
    for i in tqdm(range(100)):
        inputs = np.ones((8, 2, 2, 1)) + np.random.randn(8, 2, 2, 1)
        outputs = np.arange(8 * 2 * 2 * 3).reshape((8, 2, 2, 3))
        with tf.variable_scope('', reuse=True):
            mov_avg, mov_std, beta, gamma = sess.run([tf.get_variable('BN/moving_mean'),
                                         tf.get_variable('BN/moving_variance'),
                                                      tf.get_variable('BN/beta'),
                                                      tf.get_variable('BN/gamma')])
        if i == 0 or i == 99:
            print('\nmov_avg: {}, \nmov_std: {}, \nbeta: {}, \ngamma: {}'.format(mov_avg, mov_std, beta, gamma))
        _, _ = sess.run([train_op, tf.get_collection(tf.GraphKeys.UPDATE_OPS)], feed_dict={
            input_ph: inputs,
            output_ph: outputs,
            is_training: True,
        })

    for i in tqdm(range(100)):
        inputs = np.ones((8, 2, 2, 1)) + np.random.randn(8, 2, 2, 1)
        outputs = np.arange(8 * 2 * 2 * 3).reshape((8, 2, 2, 3))
        with tf.variable_scope('', reuse=True):
            mov_avg, mov_std, beta, gamma = sess.run([tf.get_variable('BN/moving_mean'),
                                         tf.get_variable('BN/moving_variance'),
                                                      tf.get_variable('BN/beta'),
                                                      tf.get_variable('BN/gamma')])
        if i == 0 or i == 99:
            print('\nmov_avg: {}, \nmov_std: {}, \nbeta: {}, \ngamma: {}'.format(mov_avg, mov_std, beta, gamma))
        _ = sess.run([graph.get_tensor_by_name('model/Reshape:0')], feed_dict={
            input_ph: inputs,
            is_training: False,
        })

        # print moving avg/std
    saver.save(sess, './dummy/ckpt/step100')


def freeze_ckpt_for_inference(ckpt_path=None, conserve_nodes=None):
    # clean graph first
    tf.reset_default_graph()
    # freeze ckpt then convert to pb
    new_input = tf.placeholder(tf.float32, shape=[None, 10, 10, 1], name='new_input')
    new_is_training = tf.placeholder(tf.bool, name='new_is_training')

    restorer = tf.train.import_meta_graph(
        ckpt_path + '.meta',
        input_map={
            'input_ph': new_input,
            'is_training': new_is_training,
        },
        clear_devices=True,
    )

    input_graph_def = tf.get_default_graph().as_graph_def()
    check_N_mkdir('./dummy/pb/')
    check_N_mkdir('./dummy/tb/')

    # freeze to pb
    with tf.Session() as sess:
        # restore variables
        restorer.restore(sess, './dummy/ckpt/step100')
        # convert variable to constant
        output_graph_def = tf.graph_util.convert_variables_to_constants(
            sess=sess,
            input_graph_def=input_graph_def,
            output_node_names=conserve_nodes,
        )

        # save to pb
        with tf.gfile.GFile('./dummy/pb/freeze.pb', 'wb') as f:  # 'wb' stands for write binary
            f.write(output_graph_def.SerializeToString())


def optimize_graph_for_inference(pb_dir=None, conserve_nodes=None):
    tf.reset_default_graph()
    check_N_mkdir(pb_dir)

    # import pb file
    with tf.gfile.FastGFile(pb_dir + 'freeze.pb', "rb") as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())

    # optimize graph
    optimize_graph_def = optimize_for_inference(input_graph_def=graph_def,
                                                input_node_names=['new_input', 'new_is_training'],
                                                output_node_names=conserve_nodes,
                                                placeholder_type_enum=[dtypes.float32.as_datatype_enum,
                                                                       dtypes.bool.as_datatype_enum,
                                                                       dtypes.float32.as_datatype_enum,
                                                                       ]
                           )
    with tf.gfile.GFile(pb_dir + 'optimize.pb', 'wb') as f:
        f.write(optimize_graph_def.SerializeToString())

conserve_nodes = ['model/Reshape']
freeze_ckpt_for_inference(ckpt_path='./dummy/ckpt/step100', conserve_nodes=conserve_nodes)
optimize_graph_for_inference(pb_dir='./dummy/pb/', conserve_nodes=conserve_nodes)

# cleaning
tf.reset_default_graph()

# load pb
with tf.gfile.FastGFile('./dummy/pb/optimize.pb', "rb") as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

print('\n Now inference*******************************')
# inference
with tf.Session() as sess:
    tf.graph_util.import_graph_def(
        graph_def,
    )
    # prepare
    G = tf.get_default_graph()
    print_nodes_name_shape(G)
    new_input = G.get_tensor_by_name('import/new_input:0')
    new_is_training = G.get_tensor_by_name('import/new_is_training:0')
    new_output = G.get_tensor_by_name('import/' + conserve_nodes[-1] + ':0')

    # train
    for i in tqdm(range(100)):
        inputs = np.ones((8, 2, 2, 1)) + np.random.randn(8, 2, 2, 1)
        outputs = np.arange(8 * 2 * 2 * 3).reshape((8, 2, 2, 3))
        # print moving avg/std
        with tf.variable_scope('', reuse=True):
            mov_avg, mov_std, beta, gamma = sess.run([G.get_tensor_by_name('import/BN/moving_mean:0'),
                                                      G.get_tensor_by_name('import/BN/moving_variance:0'),
                                                      G.get_tensor_by_name('import/BN/beta:0'),
                                                      G.get_tensor_by_name('import/BN/gamma:0')])
        if i == 0 or i == 99:
            print('\nmov_avg: {}, \nmov_std: {}, \nbeta: {}, \ngamma: {}'.format(mov_avg, mov_std, beta, gamma))
        new_out = sess.run([new_output], feed_dict={
            new_input: inputs,
            new_output: outputs,
            new_is_training: False,
        })
        # print('out: {}'.format(new_out))
```

Is this **an expected behavior?**. I don't think so, because the NN is very sensitive if we change even a very little bit of the beta and gamma. Is this be remedied in the 2.0 version?

**Comment by ZeliangSu** [comment]
The code is misleading the changes, since there's still one forward propagation after printing the local variables (mv_avg, mv_std, beta and gamma). But the optimize_for_inference() DOES prune the training param, since after that no-matter what you put as training (T/F), these 4 variables will not change.
